### ISSUE TITLE
Dynamic linking with tcmalloc is prone to crash

### ISSUE THREAD

**Issue by Peaker** [gperftools/gperftools#1066]
libtcmalloc, designed to override weak symbols (malloc/free), can crash when late-loaded dynamically (with `dl_open`).

Consider the code in `MallocExtension::Initialize`:
```
  string dummy("I need to be allocated");
  dummy += "!";         // so the definition of dummy isn't optimized out
```
It calls into libstdc++'s `basic_string` allocation methods, which may have already been resolved to use libc's malloc (if C++ strings were in use prior to the dl_open). Then the dtor/free calls the `delete[]` symbol directly and not via `basic_string` (probably due to inlining), which is newly loaded code, so is resolved to use `tc_free`.

This crashes with:
```
src/tcmalloc.cc:283] Attempt to free invalid pointer 0x55ef8c535fb0 
```

Using `dl_open` on ANY shared object which indirectly loads `libtcmalloc` - after instantiating `std::string` - will cause this crash.

In other words, the only safe use of `libtcmalloc` is via `-l tcmalloc` in the *MAIN* application, or via `LD_PRELOAD`. It is *not* safe to use `libtcmalloc` via `-l tcmalloc` in library code. That makes the library code dangerous to late-load.

Since late loading is a crucial technique for various programming languages (e.g: Python, Julia, Haskell in interpreted mode, ...), crashing upon late load of tcmalloc is a problem.

I don't know if there's a good fix for this, except emphasizing in the documentation that `-l tcmalloc` is only supported in the main, and will cause this terrible bug otherwise.

**Comment by obones** [comment]
I believe this is because, by default, tcmalloc will "hack" into the process and all other modules to replace malloc and free with its own implementations.
It does so when it is loaded, which is way too late when using delayed loading.
Maybe it is possible to control the "hack" in a way that it only implants itself in the current module, not in any other.

**Comment by alk** [comment]
It is not a hack or at least not supposed to be hack. tcmalloc just replaces from malloc-related symbols and thats it. I am surprised that dlopen actually causes any issues since malloc & free won't be replaced with dlopen.

Would be nice to have some more details and ideally some simple way for me to reproduce the issue.

dlopening tcmalloc is not supported, at least because we're using faster but not dl-open compatible access to thread local storage. But this kind of crash I don't get.

**Comment by thothothotho** [comment]
I have a very similar crash when trying to run a golang application that is linked to a dynamic c++ library, itself linked with -ltcmalloc:
```
src/tcmalloc.cc:283] Attempt to free invalid pointer 0x10a8190

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) where 
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff7031801 in __GI_abort () at abort.c:79
#2  0x00007ffff5527fb9 in tcmalloc::Log(tcmalloc::LogMode, char const*, int, tcmalloc::LogItem, tcmalloc::LogItem, tcmalloc::LogItem, tcmalloc::LogItem) () from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#3  0x00007ffff551bd49 in ?? () from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#4  0x00007ffff553b6b9 in tc_free () from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#5  0x00007ffff5530720 in MallocExtension::Initialize() () from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#6  0x00007ffff551ae0a in ?? () from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#7  0x00007ffff7de5733 in call_init (env=0x7fffffffdc18, argv=0x7fffffffdc08, argc=1, l=<optimized out>) at dl-init.c:72
#8  _dl_init (main_map=0x7ffff7ffe170, argc=1, argv=0x7fffffffdc08, env=0x7fffffffdc18) at dl-init.c:119
#9  0x00007ffff7dd60ca in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#10 0x0000000000000001 in ?? ()
#11 0x00007fffffffdfe0 in ?? ()
#12 0x0000000000000000 in ?? ()
```

**Comment by Peaker** [comment]
@alk I gave an explanation on how to reproduce above.

> Using `dl_open` on ANY shared object which indirectly loads `libtcmalloc` - after instantiating `std::string` - will cause this crash

This is because the linker can resolve the same symbols (`new[]`/`delete[]`) to different addresses at different points in time *in the same executable* (before/after loading tcmalloc).

If code that happened to force resolution of new/delete (e.g: basic_string module) had been loaded before tcmalloc was loaded, it will stick with the old resolution of new/delete. And so you get a mishmash of different resolutions. new/delete calls from same function can resolve differently (when one is inlined, it will re-resolve to tcmalloc, and when it calls to already-loaded-and-resolved library, it will resolve to NOT use tcmalloc).

**Comment by alk** [comment]
Yes, description above looks right. But I am not entirely sure how program fails if all new/delete operators get resolved to c++ runtime. Thus have _specific_ example I can run and see things crashing would still be nice to have. I can speculate about all kinds of ways this could happen, but knowing more about this specific case could be useful.

**Comment by obones** [comment]
Here is the situation that we have here:

We build a DLL that uses TCMalloc
That DLL gets loaded in Excel.exe via a VBA module, so this happens way after Excel.exe has been started
During DllMain, TCMalloc finds that the host uses the CRT and so injects itself into it
After that, it may take a while, but you will get issues because the memory allocated before will be deallocated by TCMalloc...

Using a build of TCMalloc that does not inject itself into the CRT solved the issue for us as we really don't need any injection into the CRT anyway.

**Comment by alk** [comment]
windows way of patching C runtime is totally unrelated to this ticket.

**Comment by obones** [comment]
Yes, but I suspect there is very similar behavior in the linux version as it also patches the C runtime "on the fly". My message was intended as an illustration of a situation where this can occur, even if it's under Windows and may appear unrelated at first.

**Comment by jtwang83** [comment]
There appears to be a dependency on gcc version. I get the crash only when compiling the shared library linking tcmalloc with g++-9.2. Using g++-7.5 to compile the shared library does not result in a crash, even if the dlopen'ing application is compiled with g++-9.2.

Inspecting the process in gdb suggests that tcmalloc does not actually load after the dlopen when compiling the .so with -ltcmalloc with g++-7.5 when main() does not link tcmalloc.

An example is provided at [https://github.com/jtwang83/tcmalloc_crash](url)

**Comment by alk** [comment]
Scrubbing. I am confused what exactly is the issue here.

Definitely no way we're going to support "dlopen case". (People mentioned right workaround above, link main program or LDPRELOAD tcmalloc)

But what is unclear is what is it, that we need to do to close this as "done" here ? People want us to somehow resist being loaded dynamically? Document it better? (isn't it already mentioned in our docs)

**Comment by beef9999** [comment]
There are tons of "Attempt to free invalid pointer" errors of tcmalloc you can searched from the Internet, but people are still asking. That means this problem is way beyond people's knowledge and expectation.

For me I was making my own shared lib libmy.so, which had statically built tcmalloc inside. Then I delivered my lib to my customers, they were also statically building tcmalloc (using same version to mine), and dynamically linking my lib. No one has ever expected to see this error, and I think this broke the basic boundary of software delivery.

**Comment by beef9999** [comment]
@alk Any advice to the issue I mentioned?

**Comment by alk** [comment]
Hi. So it is not likely that we could or should do much to deal with this case. But I am curious to see people use-cases and then I'll do my best to help.

So what was the exact intention behind linking gperftools with .so library?

**Comment by beef9999** [comment]
> Hi. So it is not likely that we could or should do much to deal with this case. But I am curious to see people use-cases and then I'll do my best to help.
> 
> So what was the exact intention behind linking gperftools with .so library?

Because our lib has conflicted symbols to our customers, we must deliver our lib by so.

And we are using tcmalloc

**Comment by alk** [comment]
My question is why you're using tcmalloc. And why can't you have your .so simply "get" tcmalloc effect from linking into main program.

**Comment by alk** [comment]
We do have several static initializers, yes. BTW one semi-anecdotal thing. I don't recall what it is, but in Debian I found one program that via one dependency got libtcmalloc.so and via other dependency got libtcmalloc_minimal.so loaded together (both via regular dynamic linking, not dlopen). And it works. Although this is clearly asking for trouble and we cannot promise it'll work in the future.

So returning to our matter. Is there anything you want us to do here? Anything we can improve?

**Comment by beef9999** [comment]
Maybe this is the root cause of all the invalid free pointers?
I think the solution is to use raw pointer + lazy initialization (call one / singleton), rather than global variable construction.

**Comment by beef9999** [comment]
I'm not aware of the dlopen thing. I was searching the error of `Attempt to free invalid pointer` and happend to see this conversation.
Now I believe this error comes from double initialization of global variable.

**Comment by alk** [comment]
Okay. So last few comments are actually unrelated to this ticket (which is purely about dlopen case). Whoever reads those comments, please keep this in mind.

As for that case described in last few comments, I am only able to speculate without seeing more details. But we actually have some evidence that multiple "duplicate" initializers are safe. One case I mentioned above. My best guess is when doing "static" linking of tcmalloc to your .so library, you end up internally calling into local copy of tcmalloc, and then returning this allocation to main program, where it gets freed by "wrong" malloc. Or vice-versa.
